.circleci/config.yml

```diff
@@ -127,57 +127,128 @@ jobs:
+
+      - restore_cache:
+          name: restore mssql installation file
+          key: downloads-{{ checksum "test/bin/install-mssql.ps1" }}
 
       - run:
-          name: build libiconv
-          no_output_timeout: 30m
+          name: setup mssql
           command: |
-            $Env:PATH = "C:\\Ruby<< parameters.ruby_version >>-x64\\bin;$Env:PATH"
-            bundle exec rake ports:libiconv
+            .\test\bin\install-mssql.ps1
+
+      - save_cache:
+          name: save downloads cache
+          paths:
+            - C:\Downloads
+          key: downloads-{{ checksum "test/bin/install-mssql.ps1" }}
 
       - run:
-          name: build freetds
-          no_output_timeout: 30m
+          name: install toxiproxy-server
           command: |
-            $Env:PATH = "C:\\Ruby<< parameters.ruby_version >>-x64\\bin;$Env:PATH"
-            bundle exec rake ports:freetds
+            choco install toxiproxy-server --version=2.5.0 -y
+            Start-Process toxiproxy-server
+
+      - restore_cache:
+          name: restore ports cache
+          keys:
+            - ports-libiconv115-openssl111s-freetds1124
 
       - run:
-          name: build gem
-          no_output_timeout: 30m
+          name: copy ports cache into project
+          command: |
+            Move-Item C:\home\circleci\project\ports .
+
+      - run:
+          name: test gem
           command: |
             $Env:PATH = "C:\\Ruby<< parameters.ruby_version >>-x64\\bin;$Env:PATH"
-            bundle exec rake ports
+            bundle exec rake test
+          environment:
+            TOXIPROXY_HOST: "localhost"
+
+  cross_compile_ports:
+    machine:
+      image: ubuntu-2004:current
+
+    steps:
+      - ruby/install:
+          version: '2.7'
+      - checkout
+      - restore_cache:
+          name: restore gem cache
+          keys:
+            - v1-bundle-{{ .Branch }}-{{ checksum "tiny_tds.gemspec" }}
+            - v1-bundle-{{ .Branch }}-
+            - v1-bundle-
+
+      - run:
+          name: bundle install gems
+          command: |
+            bundle install --path vendor/bundle
+
+      - save_cache:
+          name: save gem cache
+          paths:
+            - ./vendor/bundle
+          key: v1-bundle-{{ .Branch }}-{{ checksum "tiny_tds.gemspec" }}
+
+      - restore_cache:
+          name: restore ports cache
+          keys:
+            - ports-libiconv115-openssl111s-freetds1124
+
+      - run:
+          name: cross compile ports
+          command: |
+            bundle exec rake compile:cross
+
+      - save_cache:
+          name: save ports cache
+          paths:
+            - ./ports
+          key: ports-libiconv115-openssl111s-freetds1124
+
+      - save_cache:
+          name: save cross-compiled-code
+          paths:
+            - ./tmp
+          key: compiled-<< pipeline.git.revision >>
 
 workflows:
   test_supported_ruby_versions:
     jobs:
-      - test_linux:
-          matrix:
+      - cross_compile_ports
```

Comment:
I'm still trying to understand this strategy. If I'm understanding you, the idea is to quit trying to build the native dependencies on each OS in isolation, but rather to have a separate job here that cross compiles the native deps for Windows but from Ubuntu. On Ubuntu the strategy remains to build the native deps for itself.

If that understanding is correct, I'm conflicted about this part of the PR. The longer build time ought to be a non-issue after caching the ports folder. This also seems much more complicated but also not essential. Plus, I think it is useful to demonstrate that the native deps _can_ be compiled _on_ each OS, _for_ that OS, with _each_ version of Ruby in the list.

Another way to say it is that I see virtue in having the structure & list of job steps between the OS's (including potentially macOS in the future) be as similar as possible, even if the contents of the step needs to be altered to fit the OS (eg, different exact syntax to install/activate a particular version of Ruby, etc).

(I also recognize the counterpoint, which is the final fat gem that we publish is cross compiled, so why not do it that same way here... I'm conflicted).

Reply:
As of right now, as a Windows user I would use the fat gem while on Linux / Mac OS X I get the "regular" gem where I as the user need to make sure that I installed freetds. This is now reflected in CI with the approach I suggested.

I see the point that showing that `tiny_tds` can be compiled against those versions specified in `extconsts.conf` is helpful, although I'm not quite sure about Windows here. You can force the installation of the "regular" gem on Windows instead of the fat gem. Since `tiny_tds` specifies `msys2_mingw_dependencies` in its gemspec, freetds will be installed prior to the compile process of the gem defined with `extconf.rb`. If you want to influence the `openssl` / `libiconv` / `freetds` version in your msys2 mingw system, you need to correct this yourself.

The point I want to bring across is: If you don't use the fat gem on Windows, the compile process is out of our hands. However, the compiled ports are used for the fat gem, and it would be good if we are confident that those work.

That said, I could imagine that there could be two windows jobs, one using the precompiled code and the other one compiling the code on the system.

I added some improvements with 9d12f782. So instead of sharing the cached code between the jobs, I share the final gem. Then on Windows, I first install the appropriate fat gem, which is a nice test to also make sure the build gem works, and copy the compiled resources out of the precompiled gem into the project folder to run the tests. I think this simplifies the setup by a bit.

If you still think it's not a good idea, I'm fine with reverting the changes regarding the precompiled code and just adding the cache to the Windows jobs.

Reply:
> That said, I could imagine that there could be two windows jobs, one using the precompiled code and the other one compiling the code on the system.

Eh, I don't think that's necessary for right now

> Then on Windows, I first install the appropriate fat gem, which is a nice test to also make sure the build gem works

This makes sense and seems like a good idea

> copy the compiled resources out of the precompiled gem into the project folder to run the tests

This seems a bit weird, but maybe on balance not worth changing further right now